module/preference/PreferenceDialog.kt

```diff
@@ -0,0 +1,332 @@
+import com.vaticle.typedb.studio.state.StudioState
+import com.vaticle.typedb.studio.state.common.util.Label
+import com.vaticle.typedb.studio.state.page.Navigable
+
+object PreferenceDialog {
+    private val WIDTH = 800.dp
+    private val HEIGHT = 600.dp
+    private val NAVIGATOR_INIT_SIZE = 200.dp
+    private val NAVIGATOR_MIN_SIZE = 150.dp
+    private val STATE_INIT_SIZE = 600.dp
+    private val STATE_MIN_SIZE = 500.dp
+    private val QUERY_LIMIT_PLACEHOLDER = "1000"
+    private val IGNORED_PATHS_PLACEHOLDER = ".git, .typedb-studio"
+
+    private val appData = StudioState.appData.preferences
+    private var focusedPreferenceGroup by mutableStateOf(PreferenceGroup(""))
+
+    sealed interface Preference {
+        @Composable fun display()
+        fun valid(): Boolean
+
+        class TextInput(initial: String, var label: String, private var placeholder: String,
+                        var validator: (String) -> Boolean = { true }) : Preference {
+
+            var value by mutableStateOf(initial)
+
+            override fun valid(): Boolean {
+                return validator(value)
+            }
+
+            @Composable
+            override fun display() {
+                var border = Form.Border(1.dp, RoundedCornerShape(Theme.ROUNDED_CORNER_RADIUS)) {
+                    if (this.valid()) {
+                        Theme.studio.border
+                    } else {
+                        Theme.studio.errorStroke
+                    }
+                }
+
+                Field(label = label) {
+                    TextInput(
+                        value = value,
+                        placeholder = placeholder,
+                        border = border,
+                        onValueChange = { value = it }
+                    )
+                }
+            }
+        }
+
+        class Checkbox(initial: Boolean, var label: String): Preference {
+            var value by mutableStateOf(initial)
+            @Composable
+            override fun display() {
+                Field(label = label) {
+                    Checkbox(
+                        value = value,
+                        onChange = { value = it }
+                    )
+                }
+            }
+
+            override fun valid(): Boolean {
+                return true
+            }
+        }
+    }
+
+
+    class PreferencesForm : State {
+        // Graph Visualiser Preferences
+        var graphOutput = Preference.Checkbox(appData.graphOutput, Label.ENABLE_GRAPH_OUTPUT)
+        // Project Manager Preferences
+        val ignoredPathsData = appData.ignoredPaths.toString()
+        var ignoredPaths = Preference.TextInput(ignoredPathsData.substring(1, ignoredPathsData.length - 1),
+            Label.PROJECT_IGNORED_PATHS, IGNORED_PATHS_PLACEHOLDER)
+        // Query Runner Preferences
+        var limit = Preference.TextInput(appData.limit, Label.SET_QUERY_LIMIT, QUERY_LIMIT_PLACEHOLDER) { it.toLongOrNull() != null }
```

Comment:
This is one of those cases where the lambda syntax for the lambda parameter is just not intuitive at all.

The way I've addressed this previously is as follows (also adding named params + shortening an imported name):
```suggestion
        var limit = Preference.TextInput(value = appData.limit, label = SET_QUERY_LIMIT, placeholder = QUERY_LIMIT_PLACEHOLDER) { /* validator = */ it.toLongOrNull() != null }
```